File: crawler/internal/surfer/domain/surfing/url_generator.go
package surfing

import (
	"maps"
	"strconv"
	"sync"

	"github.com/cbroglie/mustache"
)

type URLGenerator struct {
	template *mustache.Template
}

// NewURLGenerator creates new renderer to substitute url parameters using 'mustache' syntax (logicless).
//
// https://mustache.github.io/
func NewURLGenerator(tmpl string) (URLGenerator, error) {
	parsedTmpl, err := mustache.ParseString(tmpl)
	if err != nil {
		return URLGenerator{}, err
	}
	return URLGenerator{
		template: parsedTmpl,
	}, nil
}

func (r URLGenerator) Branch(params TemplateContext) BranchURLGenerator {
	values := make(map[string]string, len(params.Values))
	maps.Copy(values, params.Values)
	return BranchURLGenerator{
		template: r.template,
		params:   values,
		mu:       &sync.Mutex{},
	}
}

type BranchURLGenerator struct {
	template *mustache.Template
	params   map[string]string

	mu *sync.Mutex
}

func (r *BranchURLGenerator) Page(pageNum int) (string, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.params[URLTemplatePageParam] = strconv.Itoa(pageNum)
	return r.template.Render(r.params)
}

File: crawler/surfer/domain/surfing/url_generator.py
import typing as tp

import chevron  # type: ignore[import-untyped]

from surfer.domain import errors, surfing


class URLGenerator:
    def __init__(self, template: str) -> None:
        try:
            chevron.render(template, {})
        except chevron.ChevronError as exc:
            raise errors.ValidationError(f"invalid mustache template: {exc}") from exc
        self._template = template

    def branch(self, ctx: surfing.TemplateContext) -> "BranchURLGenerator":
        return BranchURLGenerator(self._template, ctx.values)


class BranchURLGenerator:
    def __init__(self, template: str, params: dict[str, str]) -> None:
        self._template = template
        self._params = params

    def page(self, page_num: int) -> str:
        self._params[surfing.URL_TEMPLATE_PAGE_PARAM] = str(page_num)
        try:
            return tp.cast("str", chevron.render(self._template, self._params))
        except chevron.ChevronError as exc:
            raise errors.ValidationError(f"cannot render url: {exc}") from exc
